### ISSUE TITLE
PointNet++ gives different results running on CPU/GPU

### ISSUE THREAD

**Issue by shorohml** [rusty1s/pytorch_geometric#766]
## ❓ Questions & Help

Hello Matthias,

Thanks for awesome work on PyG. I've run into problem. I've trained PointNet++ from [example](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/pointnet2_segmentation.py) on my custom dataset. If I test model on GPU I get nice results (acc 0.97), but when I do model.to('cpu') model outputs are all 0.

Сan you tell me why this is happening?

Thanks for the help in advance.

**Comment by rusty1s** [comment]
Mh, not really sure what may cause this issue. Maybe there is a weird float/int conversion bug that converts your 0.97 float acc to 0.0 int acc?

**Comment by shorohml** [comment]
No, I've meant that model assignes class 0 to all points.

**Comment by rusty1s** [comment]
Well this is not really possible due to the softmax.

**Comment by rusty1s** [comment]
Can you verify that your parameter values stay the same after `to('cpu')`? This seems like it does a reset on these parameters.

**Comment by shorohml** [comment]
It's really strange, all parameters are the same (I verified with ==).

**Comment by shorohml** [comment]
When I'll have time, I'll try to verify it using example script on ShapeNet.

**Comment by shorohml** [comment]
I added this 3 lines at the end of [example script](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/pointnet2_segmentation.py)
![example](https://user-images.githubusercontent.com/31921963/67719512-ff243980-f9e2-11e9-852a-518007bf6553.png)
and results dropped from Acc: 0.8741, IoU: 0.7470 to Acc: 0.5689, IoU: 0.4146

**Comment by shorohml** [comment]
Should I issue this as a bug?

**Comment by rusty1s** [comment]
I look into this. This could very well be the result of a different radius_graph output.

**Comment by rusty1s** [comment]
I tracked down the issue and it is coming from the `max_num_neighbors` parameter in `pointnet2_classification`. In case the actual number of neighbors exceed this parameter, the GPU version choose the resulting neighbors randomly, and the CPU version picks the k nearest ones, resulting in different neighborhoods. This shouldn't be too much of a problem, but for the PointNet example, the actual neighborhood size is really large (can be up to 600 potential neighbors for radius size 0.2).

**Comment by shorohml** [comment]
I changed scipy.spatial.cKDTree.query call in radius to scipy.spatial.cKDTree.query_ball_tree and it started working. Not sure if it's the best solution.

**Comment by rusty1s** [comment]
Yes, this way you get all points within the radius.

**Comment by shorohml** [comment]
I mean that i get all points within the radius and then randomly pick max_num_neighbors points (like GPU version)

**Comment by rusty1s** [comment]
I added this functionality to `torch-cluster` (master branch).

**Comment by shorohml** [comment]
Ok, thanks for help.
